src/main/java/net/minecraftforge/items/IItemHandler.java

```diff
@@ -94,15 +110,131 @@
      */
     boolean isItemValid(int slot, @NotNull ItemStack stack);
+
+    /**
+     * The extraction counterpart of {@link #isItemValid(int, ItemStack)}.
+     * <p>
+     * It should be used instead of simulated extractions in cases where the contents and state of the inventory are
+     * irrelevant, mainly for the purpose of automation and logic (for instance, testing if a minecart can wait to
+     * extract items from this inventory to fill its own, or if items can never be extracted and it should move on).
+     * </p>
+     * <ul>
+     *     <li>If {@code false}, extraction of the item is never possible.</li>
+     *     <li>If {@code true}, no assumptions can be made and extraction must be simulated case-by-case.</li>
+     * </ul>
+     * The actual items in the inventory, its fullness, or any other state are <b>not</b> considered by this method.
+     *
+     * @param slot  Slot to query for validity
+     * @param stack Stack to test for validity
+     * @return {@code true} if extraction of the item from the given slot may be possible.
+     *         {@code false} if no extraction is possible under any circumstances.
+     */
+    default boolean isExtractionAllowed(int slot, @NotNull ItemStack stack)
+    {
+        return true; // TODO: 1.20 - Remove default, forcing users to explicitly choose, widening adoption
+    }
+
+    /**
+     * Queries the extended contract terms for {@link #insertItem(int, ItemStack, boolean)} and
+     * {@link #extractItem(int, int, boolean)}.
+     *
+     * @return The I/O guarantees of this item handler
+     */
+    default IOGuarantees getIOGuarantees()
+    {
+        return IOGuarantees.NONE; // TODO: 1.20 - Remove default, forcing users to explicitly choose, widening adoption
+    }
+
+    /**
+     * Enum describing the guarantees provided by insertion and extraction operations.
+     */
+    enum IOGuarantees
+    {
+        /**
+         * Insertion and extraction may affect other slots, and the results of these operations may not be reflected in
+         * the stack in the queried slot.
+         * <p>
+         * This is how item handlers have historically worked.
+         */
+        NONE(false, false),
+        /**
+         * The inventory is guaranteed to accurately reflect the outcome of insertion/extraction operations.
+         * Additionally, all slots are guaranteed to be fully independent from each other, meaning changes to one of
+         * them must not immediately affect the contents or behavior of any of the others. A delayed side effect
+         * (for example, at the end of the tick or next tick) is considered valid behavior.
+         * <p>
+         * This means that a third party may safely operate on multiple slots guaranteeing the independence of such
+         * operations.
+         */
+        CONSISTENT(true, false),
+        /**
+         * A stricter version of {@link #CONSISTENT}, providing additional guarantees on insertion and extraction.
```

Comment:
Could mention that the inventory behaves like a Container, essentially just a list of slots with insertion and extraction filters?